substrate/frame/distribution/src/functions.rs

```diff
@@ -0,0 +1,165 @@
+		spends
+	}
+
+	/// Series of checks on the Pot, to ensure that we have enough funds
+	/// before executing a Spend
+	pub fn pot_check(amount: BalanceOf<T>) -> DispatchResult {
+		// Get Pot account
+		let pot_account: AccountIdOf<T> = Self::pot_account();
+
+		// Check that the Pot as enough funds for the transfer
+		let balance = T::NativeBalance::balance(&pot_account);
+		let minimum_balance = T::NativeBalance::minimum_balance();
+		let remaining_balance = balance.saturating_sub(amount);
+
+		ensure!(remaining_balance > minimum_balance, Error::<T>::InsufficientPotReserves);
+		ensure!(balance > amount, Error::<T>::InsufficientPotReserves);
+		Ok(())
+	}
+
+	/// Funds transfer from the Pot to a project account
+	pub fn spend(
+		amount: BalanceOf<T>,
+		beneficiary: AccountIdOf<T>,
+		spend_index: u32,
+	) -> DispatchResult {
+		// Get Pot account
+		let pot_account: AccountIdOf<T> = Self::pot_account();
+
+		//Operate the transfer
+		let result =
+			T::NativeBalance::transfer(&pot_account, &beneficiary, amount, Preservation::Preserve)
+				.map_err(|_| Error::<T>::TransferFailed);
+
+		Self::process_failed_spend_result(spend_index, result)?;
+
+		Ok(())
+	}
+
+	/// Helper function used to change the status of a failed Spend
+	/// As we reserve the funds in the pot before doing a transfer
+	/// the probability of a transaction failing is very low.
+	/// However, an additionnal fail safe won't hurt.
+	pub fn process_failed_spend_result(
+		spend_index: u32,
+		result: Result<BalanceOf<T>, Error<T>>,
+	) -> Result<BalanceOf<T>, Error<T>> {
+		match result {
+			Ok(x) => {
+				// Change Spend status
+				Spends::<T>::mutate(spend_index, |val| {
+					let mut val0 = val.clone().unwrap();
+					val0.status = SpendState::Completed;
+					*val = Some(val0);
+				});
+				Ok(x)
+			},
+			Err(_e) => {
+				// Change Spend status
+				Spends::<T>::mutate(spend_index, |val| {
+					let mut val0 = val.clone().unwrap();
+					val0.status = SpendState::Failed;
+					*val = Some(val0);
+				});
+				Err(Error::<T>::FailedSpendOperation)
+			},
+		}
+	}
+
+	// Done in begin_block
+	// At the beginning of every Epoch, populate the `Spends` storage from the `Projects` storage
+	// (populated by an external process/pallet) make sure that there is enough funds before
+	// creating a new `SpendInfo`, and `ProjectInfo` corresponding to a created `SpendInfo`
+	// should be removed from the `Projects` storage. This is also a good place to Reserve the
+	// funds for created `SpendInfos`. the function will be use in a hook.
+
+	pub fn begin_block(now: BlockNumberFor<T>) -> Weight {
+		let max_block_weight = Weight::from_parts(1000_u64, 0);
+		let epoch = T::EpochDurationBlocks::get();
+
+		//We reach the check period
+		if (now % epoch).is_zero() {
+			let mut projects = Projects::<T>::get();
+
+			if projects.len() > 0 {
+				for project in projects.clone() {
+					// check if the pot has enough fund for the Spend
+					let check = Self::pot_check(project.amount);
```

Comment:
```suggestion
					let check = Self::pot_check(project.amount)?;
```

You are generally not using this syntax across your code. Let's double check you fully understand it?

Reply:
You can also do `.map_err(..)?`

Reply:
> You can also do `.map_err(..)?`

Sorry, my understanding is that `?` can only be used within functions that return `Result` or `Option`, which is not the case for `begin_block`(returning a `weight`). Is it ok to use `.expect(...)` instead?